src/NuGet.Jobs.Catalog2Registration/Hives/HiveUpdater.cs

```diff
@@ -0,0 +1,498 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using NuGet.Protocol.Catalog;
+using NuGet.Services;
+using NuGet.Services.Metadata.Catalog;
+using NuGet.Services.Metadata.Catalog.Helpers;
+using NuGet.Versioning;
+
+namespace NuGet.Jobs.Catalog2Registration
+{
+    public class HiveUpdater : IHiveUpdater
+    {
+        private static readonly IReadOnlyList<Uri> DeleteUris = new[] { Schema.DataTypes.PackageDelete };
+        
+        private readonly IHiveStorage _storage;
+        private readonly IHiveMerger _merger;
+        private readonly IEntityBuilder _entityBuilder;
+        private readonly IOptionsSnapshot<Catalog2RegistrationConfiguration> _options;
+        private readonly ILogger<HiveUpdater> _logger;
+
+        public HiveUpdater(
+            IHiveStorage storage,
+            IHiveMerger merger,
+            IEntityBuilder entityBuilder,
+            IOptionsSnapshot<Catalog2RegistrationConfiguration> options,
+            ILogger<HiveUpdater> logger)
+        {
+            _storage = storage ?? throw new ArgumentNullException(nameof(storage));
+            _merger = merger ?? throw new ArgumentNullException(nameof(merger));
+            _entityBuilder = entityBuilder ?? throw new ArgumentNullException(nameof(entityBuilder));
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public async Task UpdateAsync(
+            HiveType hive,
+            IReadOnlyList<HiveType> replicaHives,
+            string id,
+            IReadOnlyList<CatalogCommitItem> entries,
+            IReadOnlyDictionary<CatalogCommitItem, PackageDetailsCatalogLeaf> entryToCatalogLeaf,
+            CatalogCommit registrationCommit)
+        {
+            // Validate the input and put it in more convenient forms.
+            if (!entries.Any())
+            {
+                return;
+            }
+            GuardInput(entries, entryToCatalogLeaf);
+            var sortedCatalog = entries.OrderBy(x => x.PackageIdentity.Version).ToList();
+            var versionToCatalogLeaf = entryToCatalogLeaf.ToDictionary(x => x.Key.PackageIdentity.Version, x => x.Value);
+
+            // Remove SemVer 2.0.0 versions if this hive should only have SemVer 1.0.0 versions.
+            if (ShouldExcludeSemVer2(hive))
+            {
+                Guard.Assert(
+                    replicaHives.All(x => ShouldExcludeSemVer2(x)),
```

Comment:
>replicaHives.All(x => ShouldExcludeSemVer2(x)) [](start = 20, length = 46)

Super nit: `replicaHives.All(ShouldExcludeSemVer2)`. You can make line 72 `replicaHives.Any(ShouldExcludeSemVer2)` too if you'd like. I don't feel strongly about this, feel free to keep it as is if you prefer it.

Reply:
Sure, sounds fine.

Reply:
Changing to `Any` won't work because `Any` returns false if there are no elements in the sequence.